### PR TITLE
Feature/sqlalchemy

### PR DESCRIPTION
**Pull request by sampathP**
sqlalchemy implementation for masakari.

Error handling done through sqlalchemy lib and masakari only catch exceptions.
add decorators for session rollback, commit, deadlock handling.

### CODE REVIEW

**Comment by sampathP** [comment]
Conflicts are resolved.
Merge button does not show up... :imp: 
But still can merge via CLI .. :bulb:

File: masakari-controller/controller/masakari_util.py [hunk 1]
```diff
@@ -361,54 +348,42 @@ def update_notification_list_db(self, key, value, notification_id):
         :param :notification_id: Notification ID
                 (updated narrowing condition of notification list table)
         """
-        ## TODO:
-        ## Get the cursor from Func Args and remove this MySQLdb.connect
+        # TODO:
+        # Get the cursor from Func Args and remove this MySQLdb.connect
```
Comment (muroi) [inline masakari-controller/controller/masakari_util.py]: The TODO seems to be unnecessary.

File: masakari-controller/controller/masakari_worker.py [hunk 2]
```diff
@@ -570,7 +540,7 @@ def msg_filter(return_message, ignore_message_list):
         return status
 
     def _do_action_db(self, config, sql):
```
Comment (muroi) [inline masakari-controller/controller/masakari_worker.py]: When moving MySQLdb to sqlalchemy, is it the method needed?

File: masakari-controller/db/api.py [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+    with _sqlalchemy_error():
+        res = session.query(NotificationList).filter(
+            NotificationList.progress == 0,
+            NotificationList.create_at < border_time).all()
+    return res
+
+
+@_retry_on_deadlock
+@_session_handle
+def delet_expired_notification(session, progress, update_at, delete_at, id):
```
Comment (muroi) [inline masakari-controller/db/api.py]: IMO, the method don't need to have "progress" arg since following query don't use it.

**Comment by muroi** [comment]
The PR looks good to me. If possible, could you check my inline comment?

**Comment by sampathP** [comment]
@muroi Thank you for review.
Please find the fix in previous patch

**Comment by muroi** [comment]
Thank you for quick updating! I want to merge it, but I hit some following issues when I deploy this with masakari-deploy. Do you have any work for these? 
1. masakari-controller/db directory is not put under /opt/masakari/  by deb installer
2. I was required to additionally install sqlalchemy_utils ($ sudo pip install sqlalchemy_utils) even after I installed openstack

Additionally, I hit unexpected error during testing. Sorry, I should've tested it and comment it first.
1.  in recovering instance failure, I received the log in masakari-controller.log and instance is not rescued.
   <_MainThread(MainThread, started 140390764574528)> must be string, not None at  File "/opt/masakari/controller/masakari_util.py", line 205, in insert_notification_list_db
2. in recovering host failure, I received the log in masakari-controller.log and instance is not rescued.
   <Thread(Thread-4, started 139735165568768)> 'unicode' object has no attribute 'rollback' at File "/opt/masakari/controller/masakari_starter.py", line 163, in _create_vm_list_db_for_failed_host( notification_uuid)

**Comment by sampathP** [comment]
Sorry for the inconvenience.
- Add db to /opt/masakari
- I will push a patch to masakari-deploy to install sqlalchemy_utils on masakari_controller.
  Errors,
- Cause was, datetime.strptime() cant convert None to given format. 
  Fix it by checking  the exceptions of type-error, or value-error and return None if any.
  Otherwise return the datetime.strptime().
- Cause was parameter miss. parameter "session" was not given to db.api and  _session_handle() try to rollback() with the only parameter given which was  "notification_uuid".

**Comment by muroi** [comment]
Thank you @sampathP . I hit same bug you resolved by last commit.  Eventually, the PR works well. So I merged it manually since there is a merge conflict.

Close #5

**Pull request merged**